src/Domain/LeanCode.DomainModels/Ulids/Ulid.cs

```diff
@@ -0,0 +1,884 @@
+    }
+
+    public Ulid(ReadOnlySpan<byte> bytes)
+        : this()
+    {
+        if (bytes.Length != 16)
+        {
+            throw new ArgumentException("invalid bytes length, length:" + bytes.Length);
+        }
+
+        ref var src = ref MemoryMarshal.GetReference(bytes);
+        Unsafe.WriteUnaligned(ref timestamp0, Unsafe.As<byte, ulong>(ref src)); // timestamp0~randomness1
+        Unsafe.WriteUnaligned(ref randomness2, Unsafe.As<byte, ulong>(ref Unsafe.Add(ref src, 8))); // randomness2~randomness9
+    }
+
+    internal Ulid(ReadOnlySpan<char> base32)
+    {
+        // unroll-code is based on NUlid.
+
+        randomness9 = (byte)((CharToBase32[base32[24]] << 5) | CharToBase32[base32[25]]); // eliminate bounds-check of span
+
+        timestamp0 = (byte)((CharToBase32[base32[0]] << 5) | CharToBase32[base32[1]]);
+        timestamp1 = (byte)((CharToBase32[base32[2]] << 3) | (CharToBase32[base32[3]] >> 2));
+        timestamp2 = (byte)(
+            (CharToBase32[base32[3]] << 6) | (CharToBase32[base32[4]] << 1) | (CharToBase32[base32[5]] >> 4)
+        );
+        timestamp3 = (byte)((CharToBase32[base32[5]] << 4) | (CharToBase32[base32[6]] >> 1));
+        timestamp4 = (byte)(
+            (CharToBase32[base32[6]] << 7) | (CharToBase32[base32[7]] << 2) | (CharToBase32[base32[8]] >> 3)
+        );
+        timestamp5 = (byte)((CharToBase32[base32[8]] << 5) | CharToBase32[base32[9]]);
+
+        randomness0 = (byte)((CharToBase32[base32[10]] << 3) | (CharToBase32[base32[11]] >> 2));
+        randomness1 = (byte)(
+            (CharToBase32[base32[11]] << 6) | (CharToBase32[base32[12]] << 1) | (CharToBase32[base32[13]] >> 4)
+        );
+        randomness2 = (byte)((CharToBase32[base32[13]] << 4) | (CharToBase32[base32[14]] >> 1));
+        randomness3 = (byte)(
+            (CharToBase32[base32[14]] << 7) | (CharToBase32[base32[15]] << 2) | (CharToBase32[base32[16]] >> 3)
+        );
+        randomness4 = (byte)((CharToBase32[base32[16]] << 5) | CharToBase32[base32[17]]);
+        randomness5 = (byte)((CharToBase32[base32[18]] << 3) | CharToBase32[base32[19]] >> 2);
+        randomness6 = (byte)(
+            (CharToBase32[base32[19]] << 6) | (CharToBase32[base32[20]] << 1) | (CharToBase32[base32[21]] >> 4)
+        );
+        randomness7 = (byte)((CharToBase32[base32[21]] << 4) | (CharToBase32[base32[22]] >> 1));
+        randomness8 = (byte)(
+            (CharToBase32[base32[22]] << 7) | (CharToBase32[base32[23]] << 2) | (CharToBase32[base32[24]] >> 3)
+        );
+    }
+
+    // HACK: We assume the layout of a Guid is the following:
+    // Int32, Int16, Int16, Int8, Int8, Int8, Int8, Int8, Int8, Int8, Int8
+    // source: https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Private.CoreLib/src/System/Guid.cs
+    public Ulid(Guid guid)
+    {
+        Span<byte> buf = stackalloc byte[16];
+        MemoryMarshal.Write(buf, ref guid);
+        if (BitConverter.IsLittleEndian)
```

Comment:
I think `Guid.TryWriteBytes` handles some of that nonsense for you, even if that would make the code a bit slower on BE machines.